src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,157 @@
+        except Exception as e:
+            self.log.exception("{}: exception while fetching service info"
+                               .format(e))
+            pass
+        return result
+
+    def update_daemon_count(self, fs_name: str, abscount: int) -> Optional[ServiceSpec]:
+        svclist = self.get_service(fs_name)
+        if svclist is None or len(svclist) == 0:
+            self.log.warn("Failed to fetch MDS service details for fs '{}'"
+                          .format(fs_name))
+            return None
+
+        svc = svclist[0]
+
+        # no need to update placement count if the current placement count is
+        # the required count
+        if svc.spec.placement.count == abscount:
+            return None
+
+        ps = copy.deepcopy(svc.spec.placement)
+        ps.count = abscount
+        newspec = ServiceSpec(service_type=svc.spec.service_type,
+                              service_id=svc.spec.service_id,
+                              placement=ps)
+        return newspec
+
+    def get_required_standby_count(self, fs_map: dict, fs_name: str) -> int:
+        assert fs_map is not None
+        for fs in fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs['mdsmap'].get('standby_count_wanted')
+        return 0
+
+    def get_current_standby_count(self, fs_map: dict, fs_name: str, daemons: Optional[List[orchestrator.DaemonDescription]] = None) -> int:
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert fs_map is not None
+        if daemons is None:
+            return 0
+        total = 0
+        daemon_names = {
+            d.name() for d in daemons
+        }
+        for sb in fs_map['standbys']:
+            full_name = "mds.{}".format(sb['name'])
+            if full_name in daemon_names:
+                total += 1
+        return total
+
+    def get_current_active_count(self, fs_map: dict, fs_name: str) -> int:
+        assert fs_map is not None
+        for fsys in fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return len(fsys.get('mdsmap').get('up'))
+        return 0
+
+    def get_required_max_mds(self, fs_map: dict, fs_name: str) -> int:
+        assert fs_map is not None
+        for fsys in fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return fsys.get('mdsmap', {}).get('max_mds')
+        return 0
+
+    def verify_and_manage_mds_instance(self, fs_map: dict, fs_name: str):
+        assert fs_map is not None
+
+        daemons = self.get_daemons(fs_name)
+        standbys_required = self.get_required_standby_count(fs_map, fs_name)
+        standbys_current = self.get_current_standby_count(fs_map, fs_name, daemons)
+        active = self.get_current_active_count(fs_map, fs_name)
+        max_mds_required = self.get_required_max_mds(fs_map, fs_name)
```

Comment:
Log some debug information with the above values.

Reply:
fixed